aggregator/src/aggregator.rs

```diff
@@ -676,45 +679,53 @@ impl<C: Clock> Aggregator<C> {
+        // Step one: grab the existing entry for this task, if one exists. If there is no existing
+        // entry, write a new (empty) entry.
+        let cache_entry = {
+            // Unwrap safety: mutex poisoning.
+            let mut task_aggs = self.task_aggregators.lock().unwrap();
+            Arc::clone(
+                task_aggs
+                    .entry(*task_id)
+                    .or_insert_with(|| Arc::new(Mutex::default())),
+            )
+        };
 
-        // Slow path: retrieve task, create a task aggregator, store it to the cache, then return it.
-        let task_opt = self
-            .datastore
-            .run_tx("task_aggregator_get_task", |tx| {
-                let task_id = *task_id;
-                Box::pin(async move { tx.get_aggregator_task(&task_id).await })
-            })
-            .await?;
-        match task_opt {
-            Some(task) => {
-                let task_agg =
-                    Arc::new(TaskAggregator::new(task, Arc::clone(&self.report_writer))?);
-                {
-                    let mut task_aggs = self.task_aggregators.lock().await;
-                    Ok(Some(Arc::clone(
-                        task_aggs.entry(*task_id).or_insert(task_agg),
-                    )))
-                }
+        // Step two: if the entry is empty, fill it via a database query. Concurrent callers
+        // requesting the same task will contend over this lock while awaiting the result of the DB
+        // query, ensuring that in the common case only a single query will be made for each task.
+        let task_aggregator = {
+            let mut cache_entry = cache_entry.lock().await;
+            if cache_entry.is_none() {
+                *cache_entry = self
+                    .datastore
+                    .run_tx("task_aggregator_get_task", |tx| {
+                        let task_id = *task_id;
+                        Box::pin(async move { tx.get_aggregator_task(&task_id).await })
+                    })
+                    .await?
+                    .map(|task| TaskAggregator::new(task, Arc::clone(&self.report_writer)))
+                    .transpose()?
+                    .map(Arc::new);
             }
-            // Avoid caching None, in case a previously non-existent task is provisioned while the
-            // system is live. Note that for #238, if we're improving this cache to indeed cache
-            // None, we must provide some mechanism for taskprov tasks to force a cache refresh.
-            None => Ok(None),
+            cache_entry.as_ref().map(Arc::clone)
+        };
+
+        // If the task doesn't exist, remove the task entry from the cache to avoid caching a
+        // negative result. Then return the result.
+        //
+        // TODO(#238): once cache eviction is implemented, we can likely remove this step. We only
+        // need to do this to avoid trivial DoS via a requestor spraying many nonexistent task IDs.
```

Comment:
I believe we can never drop this step, because of the taskprov case--any aggregator at any time can insert a task _and_ expect it to be immediately serviceable, so we can never cache None. I think it'll ultimately be up to rate limiting to mitigate this.

Reply:
Good point, though we might decide to e.g. cache `None` in the common case & include some configuration knob to allow us to configure taskprov environments to _not_ cache `None` (`nonexistent_task_cache_duration` or something?). I updated the comment to account for taskprov's needs.

Reply:
I'll consider it for #238. I imagine it's a simple enough tweak, since we already have a flag that signals when taskprov is enabled.